1-Draft/RFCnnnn-ForEach-Parallel-Cmdlet.md

```diff
@@ -0,0 +1,99 @@
+---
+RFC: RFCnnnn
+Author: Paul Higinbotham
+Status: Draft
+SupercededBy: N/A
+Version: 1.0
+Area: Engine
+Comments Due: July 18, 2019
+Plan to implement: Yes
+---
+
+# PowerShell ForEach-Object -Parallel Cmdlet
+
+This RFC proposes a new parameter set for the existing ForEach-Object cmdlet to parallelize script block executions, instead of running them sequentially as it does now.  
+
+## Motivation
+
+    As a PowerShell User,
+    I can do simple fan-out concurrency with the PowerShell ForEach-Object cmdlet, without having to obtain and load a separate module, or deal with PowerShell jobs unless I want to.
+
+## Specification
+
+There will be two new parameter sets added to the existing ForeEach-Object cmdlet to support both synchronous and asynchronous operations for parallel script block execution.
+For the synchronous case, the `ForEach-Object` cmdlet will not return until all parallel executions complete.
+For the asynchronous case, the `ForEach-Object` cmdlet will immediately return a PowerShell job object that contains child jobs of each parallel execution.
+
+### Implementation details
+
+Implementation will be similar to the ThreadJob module.
+Script block execution will be run for each piped input on a separate thread and runspace.
+The number of threads that run at a time will be limited by a `-ThrottleLimit` parameter with a default value.
+Piped input that exceeds the allowed number of threads will be queued until a thread is available.
+For synchronous operation, a `-Timeout` parameter will be available that terminates the wait for completion after a specified time.
+Without a `-Timeout` parameter, the cmdlet will wait indefinitely for completion.
+
+### Synchronous parameter set
+
+Synchronous ForEach-Object -Parallel returns after all script blocks complete running or timeout
+
+```powershell
+ForEach-Object -Parallel -ThrottleLimit 10 -TimeoutSecs 1800 -ScriptBlock {} 
+```
+
+- `-Parallel` : parameter switch specifies fan-out parallel script block execution
+
+- `-ThrottleLimit` : parameter takes an integer value that determines the maximum number threads
+
+- `-TimeoutSecs` : parameter takes an integer that specifies the maximum time to wait for completion in seconds
+
+### Asynchronous parameter set
+
+Asynchronous ForEach-Object -Parallel immediately returns a job object for monitoring parallel script block execution
+
+```powershell
+ForEach-Object -Parallel -ThrottleLimit 5 -AsJob -ScriptBlock {}
+```
+
+- `-Parallel` : parameter switch specifies fan-out parallel script block execution
+
+- `-ThrottleLimit` : parameter takes an integer value that determines the maximum number threads
+
+- `-AsJob` : parameter switch returns a job object
+
+### Variable passing

```

Comment:
Are variables serialized and deserialized like happens when using Start-Job?  Or are references to the original object instances available in the scriptblocks?

Will scriptblocks be allowed to be passed as variables?  If so, is it within the scope of this RFC to resolve PowerShell/PowerShell#7626?  If not, will the prohibition on passing scriptblocks be enforced or merely documented?

Reply:
> Or are references to the original object instances available in the scriptblocks?

I believe it's this one. We are talking about Runspace's in the same process, so no need to serialize/deserialize.
Passing a script block as variable is not a problem. `$sb.Ast` can be passed to the target Runspace, where we call `ast.GetScriptBlock()` to get back a script block that is associated with the target Runspace.

Reply:
>...$sb.Ast can be passed to the target Runspace...

This suggests to me that ForEach-Object -Parallel would rely on users correctly converting every scriptblock to its .Ast prior to passing.  Is that the intent?  That doesn’t seem like a great strategy to me because it’s error prone.  I also don’t think that is possible to do in general.  For example, there can be modules that emit objects with scriptblock properties which are bound to a SessionState in the parent runspace.  For such modules the user would not have the opportunity to intervene.

Reply:
The current plan is to not do any serialization and to pass object copies (for value types) and references to each running scriptblock via the $using keyword and the $_ variable for piped input.  The parser already disallows assignment to a $using variable, however one can assign a reference value to another variable and assign to it, with undefined behavior.  I tried to call this out in the RFC as unsupported.

Reply:
No, the cmdlet can pull the AST off the scriptblock internally, and then pass that along.

Reply:
Yes, for scriptblock variables.  I was talking about variable types in general.